ecosystem-modules/ren/renvm-bridge/src/lib.rs

```diff
@@ -0,0 +1,159 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use codec::{Decode, Encode};
+use frame_support::{decl_error, decl_event, decl_module, decl_storage, ensure, traits::Get};
+use frame_system::{self as system, ensure_signed};
+use orml_traits::BasicCurrency;
+use primitives::Balance;
+use sp_io::{crypto::secp256k1_ecdsa_recover, hashing::keccak_256};
+use sp_runtime::DispatchResult;
+
+#[derive(Encode, Decode, Clone)]
+pub struct EcdsaSignature(pub [u8; 65]);
+
+impl PartialEq for EcdsaSignature {
+	fn eq(&self, other: &Self) -> bool {
+		&self.0[..] == &other.0[..]
+	}
+}
+
+impl sp_std::fmt::Debug for EcdsaSignature {
+	fn fmt(&self, f: &mut sp_std::fmt::Formatter<'_>) -> sp_std::fmt::Result {
+		write!(f, "EcdsaSignature({:?})", &self.0[..])
+	}
+}
+
+pub trait Trait: system::Trait {
+	type Event: From<Event<Self>> + Into<<Self as system::Trait>::Event>;
+	type Currency: BasicCurrency<Self::AccountId, Balance = Balance>;
+	/// The RenVM split public key
+	type PublicKey: Get<[u8; 20]>;
+	/// The RenVM Currency identifier
+	type CurrencyIdentifier: Get<[u8; 32]>;
+}
+
+decl_storage! {
+	trait Store for Module<T: Trait> as Template {
+		/// Signature blacklist. This is required to prevent double claim.
+		Signatures: map hasher(opaque_twox_256) EcdsaSignature => Option<()>;
+	}
+}
+
+decl_event!(
+	pub enum Event<T> where
+		<T as system::Trait>::AccountId,
+		DestAddress = [u8; 20],
+	{
+		/// Asset minted (owner, amount)
+		Minted(AccountId, Balance),
+		/// Asset burnt in this chain (owner, dest, amount)
+		Burnt(AccountId, DestAddress, Balance),
+	}
+);
+
+decl_error! {
+	pub enum Error for Module<T: Trait> {
+		/// The mint signature is invalid.
+		InvalidMintSignature,
+		/// The mint signature has already been used.
+		SignatureAlreadyUsed,
+		/// The amount being burnt is greater than the account's balance.
+		InsufficientBalance,
+	}
+}
+
+decl_module! {
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+		type Error = Error<T>;
+
+		fn deposit_event() = default;
+
+		/// Allow a user to mint if they have a valid signature from RenVM.
+		#[weight = 10_000]
+		fn mint(
+			origin,
+			p_hash: [u8; 32],
+			#[compact] amount: Balance,
+			n_hash: [u8; 32],
+			sig: EcdsaSignature,
+		) {
+			let sender = ensure_signed(origin)?;
+			Self::do_mint(sender, p_hash, amount, n_hash, sig)?;
+		}
+
+		/// Allow a user to burn assets.
+		#[weight = 10_000]
+		fn burn(
+			origin,
+			to: [u8; 20],
+			#[compact] amount: Balance,
+		) {
+			let sender = ensure_signed(origin)?;
+
+			T::Currency::withdraw(&sender, amount)?;
```

Comment:
The declared `InsufficientBalance` error is supposed to be used here?

Reply:
InsufficientBalance error removed